jscomp/config_compiler.ml

```diff
@@ -0,0 +1,129 @@
+(*
+  For Windows, we distribute a prebuilt bsc.exe
+  To build on windows, we still need figure out constructing config.ml
+  from existing compiler
+
+  For other OSes, we detect
+  if there is other compiler installed and the version matches,
+  we get the config.ml from existing OCaml compiler and build `whole_compiler`
+
+  Otherwise, we build the compiler shipped with Buckle and use the
+  old compiler.ml
+*)
+
+[@@@bs.config {no_export}]
+
+external dictOfObj : 'a Js.t -> 'b Js.Dict.t = "%identity"
+
+module Child_process = Node.Child_process
+module Process = Node.Process
+module Fs = Node.Fs
+module Path = Node.Path
+
+let delete_env_var : Process.t -> string -> unit [@bs] = [%raw{|
+  function(process, key) { delete process.env[key] }
+|}]
+
+(* need check which variables exist when we update compiler *)
+let map = [%obj {
+  _LIBDIR = "standard_library_default";
+  _BYTERUN = "standard_runtime";
+  _CCOMPTYPE = "ccomp_type";
+  _BYTECC = "bytecomp_c_compiler";
+  _BYTECCLIBS = "bytecomp_c_libraries";
+  _NATIVECC = "native_c_compiler";
+  _NATIVECCLIBS = "native_c_libraries";
+  _PACKLD = "native_pack_linker";
+  _RANLIBCMD = "ranlib";
+  _ARCMD = "ar";
+  _CC_PROFILE = "cc_profile";
+
+  _MKDLL = "mkdll"; (* undefined *)
+  _MKEXE = "mkexe"; (* undefined *)
+  _MKMAINDLL = "mkmaindll"; (* undefined TODO= upstream to print it too *)
+
+  _ARCH = "architecture";
+  _MODEL = "model";
+  _SYSTEM = "system";
+  _ASM = "asm";
+  _ASM_CFI_SUPPORTED = "asm_cfi_supported"; (* boolean *)
+  _WITH_FRAME_POINTERS = "with_frame_pointers"; (* boolean *)
+  _EXT_OBJ = "ext_obj";
+  _EXT_ASM = "ext_asm";
+  _EXT_LIB = "ext_lib";
+  _EXT_DLL = "ext_lib";
+  _HOST = "host";
+  _TARGET = "target";
+  _SYSTHREAD_SUPPORT = "systhread_supported"; (* boolean *)
+}]
+
+let patch_config jscomp_dir config_map is_windows =
+  let whole_compiler_config = Path.join [| jscomp_dir; "bin"; "config_whole_compiler.mlp" |] in
+  let whole_compiler_config_output = Path.join [| jscomp_dir; "bin"; "config_whole_compiler.ml" |] in
+  let content = Fs.readFileAsUtf8Sync whole_compiler_config in
+  let replace_values whole match_ =
+    match (match_, is_windows) with
+      | ("LIBDIR", true) ->
+        {|Filename.concat (Filename.concat (Filename.concat (Filename.dirname Sys.executable_name) "..") "lib") "ocaml"|}
+      | ("LIBDIR", false) ->
+        let origin_path = Path.join [|jscomp_dir; ".."; "lib"; "ocaml"|] in
+        Js.Json.stringify (Js.Json.string origin_path)
+      | _ ->
+        let map_val = Js.Dict.unsafeGet (dictOfObj map) match_ in
```

Comment:
`assert false` is still better than `unsafeGet`, since in the future when you do refactoring, you will get an exception rather than silent failure